corehq/apps/app_manager/util.py

```diff
@@ -507,84 +508,102 @@ def get_sort_and_sort_only_columns(detail, sort_elements):
-                            response['type'] = 'SaveToCase'
-                            form_meta['questions'].append(response)
-                        seen_save_to_case.add(question_path)
-
-
-
-
-            except XFormException as e:
-                form_meta['error'] = {
-                    'details': six.text_type(e),
-                    'edit_url': reverse(
-                        'form_source',
-                        args=[domain, app._id, form.unique_id]
-                    ),
-                }
-                form_meta['module'] = copy(module_meta)
-                errors.append(form_meta)
-            else:
-                forms.append(form_meta)
+            self.errors.append(form_meta)
+        return form_meta
+
+    def _compile_form_questions(self, form_unique_id, question):
+        from corehq.apps.reports.formdetails.readable import FormQuestionResponse
+
+        is_save_to_case = '/case/' in question['value']
+        question_path = question['value'].split('/case/')[0]
+        if is_save_to_case and question_path not in self._seen_save_to_case[form_unique_id]:
+            yield self._get_save_to_case_root_node(form_unique_id, question)
+        else:
+            response = FormQuestionResponse(question).to_json()
+            response['load_properties'] = self._case_meta.get_load_properties(form_unique_id, question['value'])
+            response['save_properties'] = self._case_meta.get_save_properties(form_unique_id, question['value'])
+            if is_save_to_case:
+                response['type'] = 'SaveToCase'
+            yield response
+
+    def _get_save_to_case_root_node(self, form_unique_id, question):
+        """Add an extra node with the root path of the save to case to attach case properties to
+        """
+        from corehq.apps.reports.formdetails.readable import FormQuestionResponse
+        question_path = question['value'].split('/case/')[0]
+        response = FormQuestionResponse({
+            "label": question_path,
+            "tag": question_path,
+            "value": question_path,
+            "repeat": question['repeat'],
+            "group": question['group'],
+            "type": 'SaveToCase',
+            "hashtagValue": question['hashtagValue'],
+            "relevant": None,
+            "required": False,
+            "comment": None,
+            "constraint": None,
+        }).to_json()
+        response['load_properties'] = self._case_meta.get_load_properties(form_unique_id, question_path)
+        response['save_properties'] = self._case_meta.get_save_properties(form_unique_id, question_path)
+        self._seen_save_to_case[form_unique_id].append(question_path)
+        return response
 
-        module_meta['forms'] = forms
-        modules.append(module_meta)
-    return modules, errors
+
+def get_form_data(domain, app, include_shadow_forms=True):
+    return FormDataGenerator(domain, app, include_shadow_forms).get_app_data()
```

Comment:
This pattern (a class that only exists for its one public method) is a little ugly, so I like that you've kept that wrapped with a simple function and constrained knowledge of the class to this file.